### PR TITLE
Увядание природы

### PR DESCRIPTION
**Pull request by Graff46**
Реализация фтчи увядания природы

### CODE REVIEW

File: src/Layers/xrRender/Blender_tree.cpp [hunk 1]
```diff
@@ -72,7 +72,7 @@ void CBlender_Tree::Compile(CBlender_Compile& C)
     }
     else
     {
-        u32	tree_aref = oNotAnTree.value ? 0 : 200;
+        u32 tree_aref = oNotAnTree.value ? 0 : 200;
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/Blender_tree.cpp:75]: Please move 'styling' changes to a separate commit to keep the history clean. A mix of style corrections and real changes makes it harder to distinguish the meaningful moments.

File: src/Layers/xrRender/Blender_Recorder_StandartBinding.cpp [hunk 2]
```diff
@@ -389,6 +389,19 @@ class cl_entity_data : public R_constant_setup //--#SM+#--
...
+    virtual void setup(R_constant* C) { RCache.set_c(C, 
+        Device.SeasonValColor[0],
+        Device.SeasonValColor[1],
+        Device.SeasonValColor[2],
+        Device.SeasonValColor[3]
+        );
+    }
+} binder_season_cb;
+
+
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/Blender_Recorder_StandartBinding.cpp:404]: Remove the extra lines please

File: src/Layers/xrRender/Blender_tree.cpp [hunk 3]
```diff
@@ -194,17 +202,18 @@ void CBlender_Tree::Compile(CBlender_Compile& C)
...
-        uber_deffer(C, false, tvs, "base", oBlend.value);
+        uber_deffer(C, false, tvs, "base", oBlend.value, nullptr, false, "clr");
         break;
     case SE_R2_SHADOW: // smap-spot
         //	TODO: DX10: Use dumb shader for shadowmap since shadows are drawn using hardware PCF
         if (oBlend.value)
             C.r_Pass(tvs_s, "shadow_direct_base_aref", FALSE, TRUE, TRUE, TRUE, D3DBLEND_ZERO, D3DBLEND_ONE, TRUE, 200);
         else
             C.r_Pass(tvs_s, "shadow_direct_base", FALSE);
+        //C.r_Constant("recolor_s", &binder_recolor_tex);
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/Blender_tree.cpp]: Remove this commented line please

File: src/xrGame/script_render_device_script.cpp [hunk 4]
```diff
@@ -40,6 +55,8 @@ SCRIPT_EXPORT(CRenderDevice, (), {
                          .def_readonly("precache_frame", &CRenderDevice::dwPrecacheFrame)
                          .def_readonly("frame", &CRenderDevice::dwFrame)
                          .def("is_paused", &is_device_paused)
-                         .def("pause", &set_device_paused),
+                         .def("pause", &set_device_paused)
+                            //Graff46 seasons
```
Comment (FreeZoneMods) [inline src/xrGame/script_render_device_script.cpp:59]: As for me - meaningless comment...

File: src/Layers/xrRender/blenders/Blender_Recorder.h [hunk 5]
```diff
@@ -176,8 +176,11 @@ class CBlender_Compile
 #endif // USE_OGL
 
     void r_ColorWriteEnable(bool cR = true, bool cG = true, bool cB = true, bool cA = true);
+    pcstr getSubPath(pcstr spt);
+    pcstr _defSubPath;
     void r_End();
 
+
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/blenders/Blender_Recorder.h:183]: Unnecessary change

File: src/Layers/xrRender/uber_deffer.cpp [hunk 6]
```diff
@@ -28,8 +28,11 @@ void uber_deffer(CBlender_Compile& C, bool hq, LPCSTR _vspec, LPCSTR _pspec, BOO
 
     string256 ps, vs, dt;
     strconcat(sizeof(vs), vs, "deffer_", _vspec, lmap ? "_lmh" : "");
-    strconcat(sizeof(ps), ps, subpath , "\\deffer_", _pspec, lmap ? "_lmh" : ""); // season
-    
+    if ((subpath) && (strlen(subpath) > 1))
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/uber_deffer.cpp:31]: It's not a good idea to use strlen to check if the string is empty. It will scan ALL the string, that's expensive. It's better to use (subpath[0] <> '\0' && subpath[1] <> '\0').

File: src/Layers/xrRender/uber_deffer.cpp [hunk 7]
```diff
@@ -28,8 +28,11 @@ void uber_deffer(CBlender_Compile& C, bool hq, LPCSTR _vspec, LPCSTR _pspec, BOO
...
+        strconcat(sizeof(ps), ps, subpath, "\\deffer_", _pspec, lmap ? "_lmh" : ""); // season
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/uber_deffer.cpp:32]: What about Linux, do we need another type of slash here?

File: src/Layers/xrRender/blenders/Blender_Recorder.h [hunk 8]
```diff
@@ -176,8 +176,11 @@ class CBlender_Compile
 #endif // USE_OGL
 
     void r_ColorWriteEnable(bool cR = true, bool cG = true, bool cB = true, bool cA = true);
+    pcstr getSubPath(pcstr spt);
+    pcstr _defSubPath;
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/blenders/Blender_Recorder.h:180]: Maybe xr_string would be better...

File: src/Layers/xrRender/blenders/Blender_Recorder.cpp [hunk 9]
```diff
@@ -357,3 +360,14 @@ void CBlender_Compile::Stage_Constant(LPCSTR name)
     int id = ParseName(name);
     passConstants.push_back(RImplementation.Resources->_CreateConstant((id >= 0) ? *lst[id] : name));
 }
+
+pcstr CBlender_Compile::getSubPath(pcstr spt)
+{
+    string_path tmp, pth;
+    strconcat(sizeof(pth), pth, GEnv.Render->getShaderPath(), spt);
+    FS.update_path(tmp, "$game_shaders$", pth);
+    if (FS.folder_exist(tmp))
```
Comment (FreeZoneMods) [inline src/Layers/xrRender/blenders/Blender_Recorder.cpp:369]: Why you want to check availability of the feature checking the existance of the folder? Maybe it will be better to use configs?

File: src/Layers/xrRender/uber_deffer.cpp (hunk 7 above)
Comment (eagleivg) [reply src/Layers/xrRender/uber_deffer.cpp:32]: Nope, FS is convert any call with full fs path

File: src/xrCore/LocatorAPI.cpp [hunk 10]
```diff
@@ -1934,6 +1931,29 @@ bool CLocatorAPI::can_modify_file(pcstr fname)
     }
 }
 
+bool CLocatorAPI::folder_exist(pcstr fullpath) // only Win ?
```
Comment (eagleivg) [inline src/xrCore/LocatorAPI.cpp:1934]: Maybe use std::filesystem::exists ?

Comment (FreeZoneMods) [reply src/xrCore/LocatorAPI.cpp:1934]: Looks like there will be problems with the files in DB-archives.

**Comment by Graff46** [comment]
Отзываю
